feat: Store and diststore for GeoRadiusByMember

azuredream opened the pull request:
- Uses `Schedule-Execute` in `GeoRadiusByMember` workflow instead of `ScheduleSingleHopT()`
- Registers command in WRITE mode instead of READ mode.
- Write if needed
#1548

azuredream commented on src/server/zset_family.cc:
```diff
@@ -3042,24 +3104,32 @@ void ZSetFamily::GeoRadiusByMember(CmdArgList args, ConnectionContext* cntx) {
     return cntx->SendError(kWrongTypeErr);
-  } else if (!result) {
+  } else if (!member_score) {
+    cntx->transaction->Conclude();
```
member_score is null and fail intermittently as mentioned here: https://github.com/dragonflydb/dragonfly/issues/1548#issuecomment-1869839316

dataset:
`geoadd Europe 13.4050 52.5200 Berlin 3.7038 40.4168 Madrid 9.1427 38.7369 Lisbon 2.3522 48.8566 Paris 16.3738 48.2082 Vienna 4.8952 52.3702 Amsterdam 10.7522 59.9139 Oslo 23.7275 37.9838 Athens 19.0402 47.4979 Budapest 6.2603 53.3498 Dublin`

Not sure if I have registered this command correctly.

romange commented on src/server/zset_family.cc:
```diff
@@ -2840,54 +2848,108 @@ void SortIfNeeded(GeoArray* ga, Sorting sorting, uint64_t count) {
+      }
+      return OpStatus::OK;
+    };
+    cntx->transaction->Execute(std::move(store_cb), false);
```
this should be `Execute(.., true)` as it's the last hop of the transaction.

azuredream replied:
When STORE is supplied, this is the last hop. Otherwise, the last hop is at line 2807.
I can either handle conclude at the end, which takes 3 lines of code, or add two more arguments, `do_store` and `already_concluded` to the whole workflow, and check them after 2807 multiple times. 
Please let me know which way do you prefer.

azuredream replied:
FYI, STORE and STOREDIST are optional:
GEORADIUSBYMEMBER key member radius <M | KM | FT | MI> [WITHCOORD]
  [WITHDIST] [WITHHASH] [COUNT count [ANY]] [ASC | DESC] [STORE key
  | STOREDIST key]

romange commented on src/server/zset_family.cc:
```diff
@@ -2840,54 +2848,108 @@ void SortIfNeeded(GeoArray* ga, Sorting sorting, uint64_t count) {
+
+    rb->SendLong(smvec.size());
+  }
+  if (write_mode) {
```
I do not think you need this conclude step in the normal flow that does not expect errors. You always know what's your last hop is and that should have `Execute(..., true)`

azuredream replied:
The last hop could vary. When STORE is not used, the last hop would be hop 2.
If I can't use conclude in this way, I'll have to add a argument "do_store" for the whole workflow.

azuredream replied:
Also, after hop 2, every conclude() would have to check do_store. If already called conclude after hop 2, skip and return.

azuredream replied:
There are at most 3 hops
hop1. Getting latitude longitude from member name
hop2. Getting nearby nodes around <latitude ,longitude >
hop3. write to store key if needed

romange replied:
Conclude is also a hop. Seem my previous comment  at line 2947

azuredream commented:
Now the code works as expected. 
Will refine the error message and add more test cases.

azuredream commented:
According to Redis' behavior, refined the error messages. Added more test cases.
Ready to be reviewed.

romange commented on src/server/zset_family.cc:
```diff
@@ -1609,7 +1612,14 @@ OpResult<unsigned> OpRem(const OpArgs& op_args, string_view key, ArgSlice member
+}
+
+OpResult<double> OpScore(const OpArgs& op_args, string_view key, string_view member,
+                         bool use_member_not_found = false) {
```
Please remove the argument `use_member_not_found`, imho, this is not a good practice to dynamically change  function behavior in such way.  Instead adapt all the callers of the function to cope with  MEMBER_NOTFOUND if needed.

romange commented on src/server/zset_family.cc:
```diff
@@ -1609,7 +1612,14 @@ OpResult<unsigned> OpRem(const OpArgs& op_args, string_view key, ArgSlice member
 }
 
-OpResult<double> OpScore(const OpArgs& op_args, string_view key, string_view member) {
+OpResult<double> OpKeyExisted(const OpArgs& op_args, string_view key) {
```
seems you never return double value from this function. something is off here.

romange commented on src/server/zset_family.cc:
```diff
@@ -2743,8 +2757,12 @@ void ZSetFamily::GeoDist(CmdArgList args, ConnectionContext* cntx) {
+//              False: Using transaction->ScheduleSingleHopT()
+OpStatus MembersOfAllNeighbors(ConnectionContext* cntx, string_view key, const GeoHashRadius& n,
+                               const GeoShape& shape_ref, GeoArray* ga, unsigned long limit,
+                               bool write_mode = false) {
```
again, changing function behavior using `write_mode` is not a good approach, imho.
Instead, I would break this function into smaller function. For example, factor out the preparation code  into a separate function that just returns `range_specs`.

romange commented on src/server/zset_family.cc:
```diff
@@ -2791,32 +2809,43 @@ bool MembersOfAllNeighbors(ConnectionContext* cntx, string_view key, const GeoHa
-    return false;
+  vector<OpResult<vector<ScoredArray>>> result_arrays;
+  if (write_mode) {
+    auto cb = [&](Transaction* t, EngineShard* shard) {
```
I would move the transaction related code to the caller (GeoSearchStoreGeneric). Basically, the division to the functions is sub-optimal imho.

romange commented on src/server/zset_family.cc:
```diff
@@ -3042,24 +3148,51 @@ void ZSetFamily::GeoRadiusByMember(CmdArgList args, ConnectionContext* cntx) {
   string_view key = ArgS(args, 0);
   // member to latlong, set shape.xy
   string_view member = ArgS(args, 1);
+  bool member_score_set = false;
```
Every Dragonfly command has the command parsing part and the execution part. We never start the execution before the parsing is complete and all the arguments are validated.
it seems you start the transaction execution before the parsing is completed. We should avoid it if possible - this just complicates the function flow.

romange reviewed (commented):
Thanks! I gave some styling/coding comments. Please take a look

azuredream commented:
Thanks for reviewing.
I've updated my PR based on your suggestions. 

Couldn't fully remove "write_mode", but it's only used in `GeoSearchStoreGeneric()`. It doesn't change function behavior a lot, just relate with using `Execute()` or `ScheduleSingleHopT()`

azuredream commented:
Hi @romange 
Please let me know if there is any question on this PR.
Thanks!

romange commented on src/server/zset_family.cc:
```diff
@@ -3032,7 +3079,49 @@ void ZSetFamily::GeoSearch(CmdArgList args, ConnectionContext* cntx) {
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    return OpKeyExisted(t->GetOpArgs(shard), key);
+  };
+  OpResult<void> result = cntx->transaction->ScheduleSingleHopT(std::move(cb));
```
I do not think the implementation of this function is correct. You use ScheduleSingleHop method on the transaction but this method is intended for operations that require only a single hop. Here you call it multiple times - which breaks the atomicity of the whole command and probably incorrect in general.   The correct way is to use Schedule() once, and then `Execute(..., false)` one or more times and finalize with Execute(...., true) or `Conclude` method in case the transaction finishes prematurely.

romange commented on src/server/zset_family.cc:
```diff
@@ -2840,54 +2820,134 @@ void SortIfNeeded(GeoArray* ga, Sorting sorting, uint64_t count) {
+//             True: Using transaction->Execute()
+//             False:  Using transaction->ScheduleSingleHopT()
+// store_key:  (Valid if geo_ops.store==GeoStoreType::kNoStore) Store result to store_key
+void GeoSearchStoreGeneric(ConnectionContext* cntx, const GeoShape& shape_ref, string_view key,
```
this function is always called on already scheduled transactions. That means it never should call ScheduleSingleHop because this function is only intended to be used for non-scheduled transactions

azuredream commented:
Updated.
That simplified the workflow a lot. 

This is my first time working on async code. Appreciate your patience!

romange commented on src/server/zset_family.cc:
```diff
@@ -2840,55 +2821,166 @@ void SortIfNeeded(GeoArray* ga, Sorting sorting, uint64_t count) {
+      }
+      return OpStatus::OK;
+    };
+    cntx->transaction->Execute(std::move(store_cb), false);
```
Again, this hop is necessary a last hop, so here you should call Execute(..., true);

romange commented on src/server/zset_family.cc:
```diff
@@ -2840,55 +2821,166 @@ void SortIfNeeded(GeoArray* ga, Sorting sorting, uint64_t count) {
+
+    rb->SendLong(smvec.size());
   }
+  cntx->transaction->Conclude();
```
you can move this call into "nostore branch"

romange commented on src/server/zset_family.cc:
```diff
@@ -2840,55 +2821,166 @@ void SortIfNeeded(GeoArray* ga, Sorting sorting, uint64_t count) {
   // if no matching results, the user gets an empty reply.
   if (ga.empty()) {
     rb->SendNull();
+    cntx->transaction->Conclude();
```
what's the expected behaviour for  no matched result together with STORE option? do we have a test for that? are we compliant with redis behavior in this case?

azuredream commented:
Updated.

Added one more test case where no matching result in the search area. This can only happen in `GEOSEARCH FROMLONGLAT` because searching by member always has at least one result(the member itself). 
Redis returns an empty array in this case, so I updated the code to do the same.

romange reviewed (approved):
Thank you very much for progressing with this!

azuredream commented:
You are very welcome. 
Please let me know if you need GEOSEARCHSTORE as well.

BTW, if you are planning to implement vector related functions, I'll be glad to help.
